dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/ingest/UningestService.java

```diff
@@ -0,0 +1,189 @@
+        this.mapLayerMetadataService = mapLayerMetadataService;
+        this.datasetVersionRepository = datasetVersionRepository;
+        this.dataTableRepository = dataTableRepository;
+        this.ingestReportRepository = ingestReportRepository;
+        this.dataFileTagRepository = dataFileTagRepository;
+        this.dataFileRepository = dataFileRepository;
+        this.fileMetadataRepository = fileMetadataRepository;
+        this.datasetVersionService = datasetVersionService;
+    }
+
+    // -------------------- LOGIC --------------------
+
+    public void uningest(List<DataFile> files, AuthenticatedUser user) {
+        if (files == null || files.isEmpty() || user == null) {
+            return;
+        }
+        DataFile anyMergedFile = null;
+        for (DataFile file : files) {
+            anyMergedFile = uningestFile(file, user);
+        }
+        // As all files are from the same version it's sufficient to use
+        // any of the files to recalculate UNF
+        recalculateUNFForCurrentVersion(anyMergedFile);
+    }
+
+    // -------------------- PRIVATE --------------------
+
+    private DataFile uningestFile(DataFile file, AuthenticatedUser user) {
+        // File to uningest cannot be already present in any published version
+        if (file.getFileMetadatas().size() > 1 || !file.getFileMetadata().getDatasetVersion().isDraft()) {
+            throw new IllegalArgumentException("Only fresh files from draft can be uningested!");
+        }
+        // Variable file is updated in order to avoid OptimisticLockException
+        file = file.isTabularData()
+                ? uningestTabularFile(file)
+                : uningestFileWithoutDataTable(file);
+        return removeWorldMapData(file, user);
+    }
+
+    private DataFile uningestTabularFile(DataFile file) {
+        long originalSize;
+        StorageIO<DataFile> storage;
+        try {
+            storage = dataAccess.getStorageIO(file);
+            storage.open();
+            originalSize = storage.getAuxObjectSize(StorageIOConstants.SAVED_ORIGINAL_FILENAME_EXTENSION);
+            storage.revertBackupAsAux(StorageIOConstants.SAVED_ORIGINAL_FILENAME_EXTENSION);
+        } catch (IOException ioe) {
+            logger.warn("Uningest problem, rethrowing: ", ioe);
+            throw new RuntimeException(ioe);
+        }
+
+        // Restore size & format
+        file.setFilesize(originalSize);
+        String originalFormat = file.getDataTable().getOriginalFileFormat();
+        file.setContentType(originalFormat);
+
+        // Remove data table
+        DataTable dataTable = file.getDataTable();
+        dataTableRepository.deleteById(dataTable.getId());
+        file.setDataTable(null);
+
+        // Remove tags (only properly ingested file can have these tags)
+        dataFileTagRepository.removeByIds(file.getTags().stream()
+                .map(DataFileTag::getId)
+                .collect(Collectors.toList()));
+        file.setTags(null);
```

Comment:
I have a hard time to understand why we are not rely on jpa cascading mechanism. From what I can understand.
Making:
```
file.setDataTable(null);
file.setTags(null);
```
Should be all we need (without separate `dataTableRepository.deleteById` and `dataFileTagRepository.removeByIds`) since `DataTable` and `DataFileTag` cascading is turned on.

Reply:
Cascading is not sufficient – also orphan removal must be explicitly turned on.
On tags it's ok, I've also added that on ingest reports. However with data table it would require supplying annotations on all dependent entities – so I left that as it was. 
Nevertheless, two of repositories have been removed.